inc/knowbaseitem.class.php

```diff
@@ -399,99 +399,55 @@ public function haveVisibilityAccess() {
-              && isset($_SESSION["glpiactiveprofile"]['id']))) {
-         $join .= " LEFT JOIN `glpi_knowbaseitems_profiles`
-                        ON (`glpi_knowbaseitems_profiles`.`knowbaseitems_id`
-                              = `glpi_knowbaseitems`.`id`) ";
-      }
-
-      // Entities
-      if ($forceall
-          || !Session::getLoginUserID()
-          || (isset($_SESSION["glpiactiveentities"]) && count($_SESSION["glpiactiveentities"]))) {
-         $join .= " LEFT JOIN `glpi_entities_knowbaseitems`
-                        ON (`glpi_entities_knowbaseitems`.`knowbaseitems_id`
-                              = `glpi_knowbaseitems`.`id`) ";
-      }
-
-      return $join;
+      $it = new \DBmysqlIterator(null);
+      $it->execute($criteria);
+      $sql = $it->getSql();
+      $sql = str_replace(
+         'SELECT * FROM '.$DB->quoteName(self::getTable()).' ',
+         '',
+         $sql
+      );
+      return $sql;
    }
 
    /**
     * Return visibility SQL restriction to add
     *
     * @since 0.83
+    * @deprecated 9.4.0
     *
     * @return string restrict to add
    **/
    static function addVisibilityRestrict() {
+      Toolbox::deprecated('Use getVisibilityCriteria');
 
-      if (Session::haveRight(self::$rightname, self::KNOWBASEADMIN)) {
-         return "1";
-      }
-
-      $restrict = '';
-      if (Session::getLoginUserID()) {
-         $restrict = "(`glpi_knowbaseitems`.`users_id` = '".Session::getLoginUserID()."' ";
-
-         // Users
-         $restrict .= " OR `glpi_knowbaseitems_users`.`users_id` = '".Session::getLoginUserID()."' ";
-
-         // Groups
-         if (isset($_SESSION["glpigroups"]) && count($_SESSION["glpigroups"])) {
-            $restrict .= " OR (`glpi_groups_knowbaseitems`.`groups_id`
-                                    IN ('".implode("','", $_SESSION["glpigroups"])."')
-                               AND (`glpi_groups_knowbaseitems`.`entities_id` < 0
-                                    ".getEntitiesRestrictRequest("OR", "glpi_groups_knowbaseitems",
-                                                                 '', '', true).")) ";
-         }
+      //get and clean criteria
+      $criteria = self::getVisibilityCriteria();
+      unset($criteria['LEFT JOIN']);
+      $criteria['FROM'] = self::getTable();
 
-         // Profiles
-         if (isset($_SESSION["glpiactiveprofile"])
-             && isset($_SESSION["glpiactiveprofile"]['id'])) {
-            $restrict .= " OR (`glpi_knowbaseitems_profiles`.`profiles_id`
-                                    = '".$_SESSION["glpiactiveprofile"]['id']."'
-                               AND (`glpi_knowbaseitems_profiles`.`entities_id` < 0
-                                    ".getEntitiesRestrictRequest("OR", "glpi_knowbaseitems_profiles",
-                                                                 '', '', true).")) ";
-         }
+      $it = new \DBmysqlIterator(null);
+      $it->execute($criteria);
```

Comment:
This query should just be evaluated, not executed.
```suggestion
      $it->buildQuery($criteria);
```